src/PowerShellEditorServices/Services/PowerShell/Host/InternalHost.cs

```diff
@@ -0,0 +1,778 @@
+
+            try
+            {
+                if (_psFrameStack.Count == 1)
+                {
+                    RunTopLevelExecutionLoop();
+                }
+                else if ((frame.FrameType & PowerShellFrameType.Debug) != 0)
+                {
+                    RunDebugExecutionLoop();
+                }
+                else
+                {
+                    RunExecutionLoop();
+                }
+            }
+            finally
+            {
+                PopPowerShell();
+            }
+        }
+
+        private void PopPowerShell()
+        {
+            _shouldExit = false;
+            PowerShellContextFrame frame = _psFrameStack.Pop();
+            try
+            {
+                // If we're changing runspace, make sure we move the handlers over
+                if (_runspaceStack.Peek().Item1 != CurrentPowerShell.Runspace)
+                {
+                    (Runspace parentRunspace, _) = _runspaceStack.Pop();
+                    RemoveRunspaceEventHandlers(frame.PowerShell.Runspace);
+                    AddRunspaceEventHandlers(parentRunspace);
+                }
+            }
+            finally
+            {
+                frame.Dispose();
+            }
+        }
+
+        private void RunTopLevelExecutionLoop()
+        {
+            // Make sure we execute any startup tasks first
+            if (_psFrameStack.Count == 1)
+            {
+                while (_taskQueue.TryTake(out ISynchronousTask task))
+                {
+                    task.ExecuteSynchronously(CancellationToken.None);
+                }
+            }
+
+            RunExecutionLoop();
+        }
+
+        private void RunDebugExecutionLoop()
+        {
+            try
+            {
+                DebugContext.EnterDebugLoop(CancellationToken.None);
+                RunExecutionLoop();
+            }
+            finally
+            {
+                DebugContext.ExitDebugLoop();
+            }
+        }
+
+        private void RunExecutionLoop()
+        {
+            while (!_shouldExit)
+            {
+                using (CancellationScope cancellationScope = _cancellationContext.EnterScope(isIdleScope: false))
+                {
+                    DoOneRepl(cancellationScope.CancellationToken);
+
+                    if (_shouldExit)
+                    {
+                        break;
+                    }
+
+                    while (!cancellationScope.CancellationToken.IsCancellationRequested
+                        && _taskQueue.TryTake(out ISynchronousTask task))
+                    {
+                        task.ExecuteSynchronously(cancellationScope.CancellationToken);
+                    }
+                }
+            }
+        }
+
+        private void DoOneRepl(CancellationToken cancellationToken)
+        {
+            if (_skipNextPrompt)
+            {
+                _skipNextPrompt = false;
+                return;
+            }
```

Comment:
We need this so that when we run something that cancels the current task to run immediately, the REPL doesn't get in its way